packages/brokers/src/brokers/redis/BaseRedis.ts

```diff
@@ -0,0 +1,128 @@
+import type { Buffer } from 'node:buffer';
+import { readFileSync } from 'node:fs';
+import { resolve } from 'node:path';
+import { AsyncEventEmitter } from '@vladfrangu/async_event_emitter';
+import type { Redis } from 'ioredis';
+import { ReplyError } from 'ioredis';
+import type { BaseBrokerOptions, IBaseBroker, ToEventMap } from '../Broker.js';
+import { DefaultBrokerOptions } from '../Broker.js';
+
+// For some reason ioredis doesn't have this typed, but it exists
+declare module 'ioredis' {
+	interface Redis {
+		xreadgroupBuffer(...args: (Buffer | string)[]): Promise<[Buffer, [Buffer, Buffer[]][]][] | null>;
+	}
+}
+
+export interface RedisBrokerOptions extends BaseBrokerOptions {
+	redisClient: Redis;
+}
+
+export abstract class BaseRedisBroker<TEvents extends Record<string, any>>
+	extends AsyncEventEmitter<ToEventMap<TEvents>>
+	implements IBaseBroker<TEvents>
+{
+	public static readonly STREAM_DATA_KEY = 'data';
+
+	protected readonly options: Required<RedisBrokerOptions>;
+
+	protected readonly subscribedEvents = new Set<string>();
+
+	protected readonly streamReadClient: Redis;
+
+	protected listening = false;
+
+	public constructor(options: RedisBrokerOptions) {
+		super();
+		this.options = { ...DefaultBrokerOptions, ...options };
+		options.redisClient.defineCommand('xcleangroup', {
+			numberOfKeys: 1,
+			lua: readFileSync(resolve(__dirname, '..', '..', '..', 'scripts', 'xcleangroup.lua'), 'utf8'),
+		});
+		this.streamReadClient = options.redisClient.duplicate();
+	}
+
+	public async subscribe(group: string, events: (keyof TEvents)[]): Promise<void> {
+		await Promise.all(
+			// eslint-disable-next-line consistent-return
+			events.map(async (event) => {
+				this.subscribedEvents.add(event as string);
+				try {
+					return await this.options.redisClient.xgroup('CREATE', event as string, group, 0, 'MKSTREAM');
```

Comment:
Is the return necessary?

Reply:
I suppose not. It just seemed intuitive to me given it's inside a `.map` call fed to `Promise.all`, but the behavior would indeed be the same without it.